ibtk/src/utilities/RestartCleaner.cpp

```diff
@@ -0,0 +1,269 @@
+    {
+        return CleanupStrategy::KEEP_RECENT_N;
+    }
+
+    TBOX_ERROR(d_object_name << "::parseStrategy(): Unknown strategy: " << strategy_str << std::endl);
+    return CleanupStrategy::KEEP_RECENT_N; // Never reached
+} // parseStrategy
+
+void
+RestartCleaner::executeStrategy() const
+{
+    switch (d_strategy)
+    {
+    case CleanupStrategy::KEEP_RECENT_N:
+        keepRecentN();
+        break;
+    default:
+        TBOX_ERROR(d_object_name << "::executeStrategy(): Unknown strategy" << std::endl);
+    }
+} // executeStrategy
+
+int
+RestartCleaner::parseIterationNum(const std::string& dirname) const
+{
+    std::regex pattern("restore\\.([0-9]{6})");
+    std::smatch match;
+
+    if (std::regex_match(dirname, match, pattern))
+    {
+        return std::stoi(match[1].str());
+    }
+    return -1;
+} // parseIterationNum
+
+std::vector<std::filesystem::path>
+RestartCleaner::getAllRestartDirs(const std::string& restart_dir) const
+{
+    std::vector<std::filesystem::path> restart_dirs;
+
+    if (!std::filesystem::exists(restart_dir))
+    {
+        TBOX_ERROR(d_object_name << "::getAllRestartDirs(): Directory does not exist: " << restart_dir << std::endl);
+    }
+
+    if (!std::filesystem::is_directory(restart_dir))
+    {
+        TBOX_ERROR(d_object_name << "::getAllRestartDirs(): Path is not a directory: " << restart_dir << std::endl);
+    }
+
+    std::regex pattern("restore\\.([0-9]{6})");
+
+    try
+    {
+        for (const auto& entry : std::filesystem::directory_iterator(restart_dir))
+        {
+            if (entry.is_directory())
+            {
+                std::string dirname = entry.path().filename().string();
+                if (std::regex_match(dirname, pattern))
+                {
+                    restart_dirs.push_back(entry.path());
+                }
+            }
+        }
+    }
+    catch (const std::filesystem::filesystem_error& ex)
+    {
+        TBOX_ERROR(d_object_name << "::getAllRestartDirs(): Filesystem error: " << ex.what() << std::endl);
+    }
+
+    return restart_dirs;
+} // getAllRestartDirs
+
+void
+RestartCleaner::keepRecentN() const
+{
+    auto dirs = getAllRestartDirs(d_restart_base_path);
+
+    if (dirs.size() <= static_cast<size_t>(d_keep_restart_count))
+    {
+        if (d_log_actions)
+        {
+            plog << d_object_name << "::keepRecentN(): Found " << dirs.size()
+                 << " directories, keeping all (threshold: " << d_keep_restart_count << ")" << std::endl;
+        }
+        return;
+    }
+
+    // Sort directories by iteration number (ascending order)
+    std::sort(dirs.begin(),
+              dirs.end(),
+              [this](const std::filesystem::path& a, const std::filesystem::path& b)
+              {
+                  int iter_a = parseIterationNum(a.filename().string());
+                  int iter_b = parseIterationNum(b.filename().string());
+                  return iter_a < iter_b;
+              });
```

Comment:
we need to correctly filter-out erroneously named directories here. One possibility would be to use `std::partition()`, with the first half having invalid values for iteration number and the second half having valid values.